### PR TITLE
[BEAM-9252] Exclude jboss's Main and module-info.java

### PR DESCRIPTION
**Pull request by suztomo**
CC: @robinyqiu 

# How to install the vendored JAR locally
```
suztomo@suxtomo24:~/beam6/vendor/grpc-1_26_0$ ../../gradlew -PvendoredDependenciesOnly -Ppublishing publishToMavenLocal
Configuration on demand is an incubating feature.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 12s
4 actionable tasks: 4 executed
suztomo@suxtomo24:~/beam6/vendor/grpc-1_26_0$ jar tf /usr/local/google/home/suztomo/.m2/repository/org/apache/beam/beam-vendor-grpc-1_26_0/0.2-SNAPSHOT/beam-vendor-grpc-1_26_0-0.2-SNAPSHOT.jar |grep org/jboss/modules/Main
suztomo@suxtomo24:~/beam6/vendor/grpc-1_26_0$ jar tf /usr/local/google/home/suztomo/.m2/repository/org/apache/beam/beam-vendor-grpc-1_26_0/0.2-SNAPSHOT/beam-vendor-grpc-1_26_0-0.2-SNAPSHOT.jar |grep module-info
suztomo@suxtomo24:~/beam6/vendor/grpc-1_26_0$ 
```

# How to use the vendored JAR from Maven project

```
suztomo@suxtomo24:~/jbonofre-beam-samples$ git diff
diff --git a/pom.xml b/pom.xml
index 6296cf2..2a366e1 100644
--- a/pom.xml
+++ b/pom.xml
@@ -79,6 +79,12 @@
 
     <dependencyManagement>
         <dependencies>
+            <!-- Temporary override to confirm the shading issue (BEAM-9252) -->
+            <dependency>
+                <groupId>org.apache.beam</groupId>
+                <artifactId>beam-vendor-grpc-1_26_0</artifactId>
+                <version>0.2-SNAPSHOT</version>
+            </dependency>
```

The solution worked for jbonofre-beam-samples shading: https://gist.github.com/suztomo/929f82458c30176de669e0abf1a443a4 .

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by suztomo** [comment]
@robinyqiu I appreciate if you can confirm that this fix works in Cloud Dataflow. I only confirmed shading plugin works.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/VendorJavaPlugin.groovy [hunk 1]
```diff
@@ -131,15 +131,15 @@ artifactId=${project.name}
         inputs.files project.configurations.shadow.artifacts.files
         doLast {
           project.configurations.shadow.artifacts.files.each {
-            FileTree exposedClasses = project.zipTree(it).matching {
+            FileTree unexpectedlyExposedClasses = project.zipTree(it).matching {
               include "**/*.class"
               exclude "org/apache/beam/vendor/**"
               // BEAM-5919: Exclude paths for Java 9 multi-release jars.
-              exclude "**/module-info.class"
```
Comment (lukecwik) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/VendorJavaPlugin.groovy]: A similar change is needed in:
https://github.com/apache/beam/blob/1133f7b5ae5d10120563e267d7c65f669ccc2bf1/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L970

Comment (lukecwik) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/VendorJavaPlugin.groovy]: The glob matching semantics [allegedly](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/AbstractCopyTask.html#include-java.lang.String...-) follow [ANT pattern matching](https://ant.apache.org/manual/dirtasks.html#patterns) so I'm lost as to why `**/module-info.class` didn't exclude the top level pattern.

Comment (robinyqiu) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/VendorJavaPlugin.groovy]: `**/module-info.class` is excluding the top level pattern. If I understand it correctly, the problem is we only want the top level one to be excluded. This was a regression introduced [here](https://github.com/apache/beam/pull/10463/files#diff-23833058cbf2c1172b90e7764032aa59R959) during the last grpc version change.

Comment (lukecwik) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/VendorJavaPlugin.groovy]: Now I understand, misread what the exclusion was for. Thought it was for excluding the file and not for excluding files during validation.

Comment (suztomo) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/VendorJavaPlugin.groovy]: Done

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 2]
```diff
@@ -967,7 +967,7 @@ class BeamModulePlugin implements Plugin<Project> {
                 FileTree exposedClasses = project.zipTree(it).matching {
                   include "**/*.class"
                   // BEAM-5919: Exclude paths for Java 9 multi-release jars.
-                  exclude "**/module-info.class"
+                  exclude "META-INF/versions/*/module-info.class"
```
Comment (suztomo) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Reverting this change https://github.com/apache/beam/pull/10463/files#r382787180 . validateShadedJarDoesntLeakNonProjectClasses should fail upon root module-info.class.

**Comment by lukecwik** [comment]
Lets hear back from @robinyqiu before merging.

**Comment by robinyqiu** [comment]
LGTM. (Verified this is working on Flink and Dataflow runner.)

**Comment by suztomo** [comment]
@robinyqiu Thank you.

@lukecwik Would you take care of the release of this vendor grpc change?

**Pull request merged**